share/wake/lib/core/list.wake

```diff
@@ -15,219 +15,455 @@
-export def head = match _
-  Nil  = None
-  h, _ = Some h
-
-# Returns a list of all elements, starting from the second.
-export def tail = match _
-  Nil  = Nil
-  _, t = t
-
-# Applies a function f to each element of the list.
-export def map f =
-  def helper = match _
-    Nil  = Nil
-    h, t = f h, helper t
-  helper
+    Nil = True
+    _   = False
 
-# Applies a function to each element of the List and builds a new List from the resulting elements
-# (f: a => List b) => List a => List b
-export def mapFlat f =
-  def helper = match _
-    Nil  = Nil
-    h, t = f h ++ helper t
-  helper
+# head: retrieve the first element of the list, else None.
+# If you find yourself using the function, consider using match instead.
+#
+#   head Nil = None
+#   head ("a", "b", Nil) = Some "a"
+#   head (seq 10) = Some 0
+export def head: List a => Option a = match _
+    Nil  = None
+    h, _ = Some h
 
-export def mapPartial f =
-  def helper = match _
+# tail: remove the first element from the List
+# If you find yourself using the function, consider using match instead.
+#
+#   tail (seq 5) = 1, 2, 3, 4, Nil
+#   tail ("a", Nil) = Nil
+#   tail Nil = Nil
+#   tail (pi, 1.0, Nil) = 1.0
+export def tail: List a => List a = match _
     Nil  = Nil
-    h, t =
-      def sub = helper t
-      match (f h)
-        Some x = x, sub
-        None   = sub
-  helper
-
-export def foldl f =
-  def helper a = match _
-    Nil  = a
-    h, t = helper (f a h) t
-  helper
-
-export def scanl f =
-  def helper a = match _
-    Nil  = a, Nil
-    h, t = a, helper (f a h) t
-  helper
-
-export def foldr f =
-  def helper a = match _
-    Nil  = a
-    h, t = f h (helper a t)
-  helper
-
-export def scanr f =
-  def helper a = match _
-    Nil  = Pair a (a, Nil)
-    h, t =
-      def Pair u v = helper a t
-      def head = f h u
-      Pair head (head, v)
-  helper _ _ | getPairSecond
-
-# list helpers
-def lcat = foldr (_,_) # curried to expose recursive helper outside of ++
-export def l ++ r  = lcat r l # ++ is required by the implementation of publish
-export def reverse = foldl (\a\b b,a) Nil
-export def flatten = foldr (_++_) Nil
-export def len     = foldl (\a\_ a+1) 0
-
-# Place item at start of list. This is O(1).
-export def prepend item list = item, list
-
-# Place item at end of list. This is O(n), use prepend or ',' for O(1).
-export def append  item list = list ++ (item, Nil)
-
-# list choppers
-export def splitAt i l =
-  if i <= 0 then Pair Nil l else match l
-    Nil  = Pair Nil Nil
-    h, t =
-      def Pair u v = splitAt (i-1) t
-      Pair (h, u) v
-
-export def take i l =
-  if i <= 0 then Nil else match l
-    Nil = Nil
-    h, t = h, take (i-1) t
+    _, t = t
 
-export def drop i l =
-  if i <= 0 then l else match l
-    Nil = Nil
-    _, t = drop (i-1) t
+# map: create a new List by applying a function f to each element of a List.
+#
+#   map str    (seq 5) = "0", "1", "2", "3", "4", Nil
+#   map (_+10) (seq 5) = 10, 11, 12, 13, 14, Nil
```

Comment:
In some examples there is a `---` to separate the wake code from the result of computation
It could be nice to have a consistent way to communicate this

Reply:
I agree, though I would like this decision to be guided by whatever the tooling we use expects.